File: retired/ts-backend/central-api/src/modules/identity/identity.service.ts
import { Inject, Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtVerificationError } from './jwt/jwt-errors';
import {
  IDENTITY_PROVIDER,
  type AuthPrincipal,
  type IdentityProviderAdapter,
  type ProviderMetadata,
  type ValidatedToken,
} from './identity-provider.adapter';

@Injectable()
export class IdentityService {
  constructor(@Inject(IDENTITY_PROVIDER) private readonly provider: IdentityProviderAdapter) {}

  getProvider(): ProviderMetadata {
    return this.provider.getProviderMetadata();
  }

  async validateToken(accessToken: string): Promise<ValidatedToken> {
    try {
      return await this.provider.validateToken(accessToken);
    } catch (error) {
      if (error instanceof JwtVerificationError) {
        throw new UnauthorizedException(`Invalid access token: ${error.reason}`);
      }
      throw error;
    }
  }

  async resolvePrincipal(accessToken: string): Promise<AuthPrincipal> {
    const validated = await this.validateToken(accessToken);
    const identity = await this.provider.getUserIdentity(validated);
    const user = await this.provider.mapExternalUser(identity);

    if (!user) {
      throw new UnauthorizedException('Access token is valid but the user is not mapped to the platform');
    }

    return { user, identity };
  }
}

File: backend/central-api/src/main/java/com/cybelinx/platform/api/security/IdentityService.java
package com.cybelinx.platform.api.security;

import com.cybelinx.platform.api.common.error.ApiHttpException;
import com.cybelinx.platform.api.security.identity.IdentityClaims;
import com.cybelinx.platform.api.security.identity.IdentityProvider;
import com.cybelinx.platform.api.security.identity.IdentityToken;
import com.cybelinx.platform.api.security.identity.IdentityVerificationException;
import java.util.Map;

/**
 * Resolves bearer tokens to platform principals via the configured {@link IdentityProvider}.
 * Authentication is delegated; the platform handles user mapping and downstream authorization.
 */
public final class IdentityService {

    private final IdentityProvider provider;
    private final UserMappingService userMapping;

    public IdentityService(IdentityProvider provider, UserMappingService userMapping) {
        this.provider = provider;
        this.userMapping = userMapping;
    }

    public ProviderMetadata getProviderMetadata() {
        IdentityProvider.Metadata metadata = provider.metadata();
        return new ProviderMetadata(
                metadata.provider(),
                metadata.issuer(),
                metadata.audience(),
                metadata.supportedAlgorithms(),
                metadata.supportsJwks());
    }

    public ValidatedToken validateToken(String accessToken) {
        try {
            IdentityToken token = provider.verify(accessToken);
            return new ValidatedToken(
                    token.rawToken(),
                    provider.metadata().provider(),
                    token.algorithm(),
                    token.keyId(),
                    token.claims().raw());
        } catch (IdentityVerificationException error) {
            throw ApiHttpException.unauthorized("Invalid access token: " + error.reason().name());
        }
    }

    public AuthPrincipal resolvePrincipal(String accessToken) {
        ValidatedToken validated = validateToken(accessToken);
        AuthPrincipal.AuthIdentity identity = toAuthIdentity(validated);
        AuthPrincipal.AuthUser user = userMapping.resolveUser(identity);

        if (user == null) {
            throw ApiHttpException.unauthorized("Access token is valid but the user is not mapped to the platform");
        }

        return new AuthPrincipal(user, identity);
    }

    private AuthPrincipal.AuthIdentity toAuthIdentity(ValidatedToken validated) {
        Map<String, Object> claims = validated.claims();
        IdentityClaims parsed = IdentityClaims.from(claims);
        return new AuthPrincipal.AuthIdentity(
                provider.metadata().provider(),
                parsed.subject(),
                parsed.email(),
                parsed.name());
    }

    public record ProviderMetadata(
            String provider, String issuer, java.util.List<String> audience,
            java.util.List<String> supportedAlgorithms, boolean supportsJwks) {}

    public record ValidatedToken(String token, String provider, String alg, String kid, Map<String, Object> claims) {}
}